### DISCUSSION TITLE
[HELP] Struggling to make custom widget implement Scrollable

### DISCUSSION THREAD

**Discussion by LittleHuba** [gtk-rs/gtk4-rs#1922] [Q&A]
I have the following attempt to subclass a widget with some interfaces. One of those interfaces is `Scrollable`.

```
glib::wrapper! {
    pub struct CustomViewport(ObjectSubclass<imp::CustomViewport>)
    @extends Widget,
    @implements Accessible, Buildable, ConstraintTarget, Scrollable;
}
```

But if I check in the GTK Inspector what interfaces my widget implements, I only see `Accessible, Buildable, ConstraintTarget` but not `Scrollable`.

Is there something I'm doing wrong?

**Comment by bilelmoussaoui** [comment]
The `@implements` part of glib::wrapper! only implements the IsA<T> traits. If you want your type to be a Scrollable, you have to write something like `impl ScrollableImpl for CustomViewport {}` in the imp module and providing the required vfuncs. You would also need to override certain properties, see https://gitlab.gnome.org/World/obfuscate/-/blob/master/src/widgets/drawing_area.rs?ref_type=heads#L54 for a full example
